src/native/runtime-base/monodroid-dl.hh

```diff
@@ -29,18 +29,65 @@ namespace xamarin::android::internal
 			return lflags;
 		}
 
-		static DSOCacheEntry* find_dso_cache_entry (hash_t hash) noexcept
+		template<bool AotCache>
+		[[gnu::always_inline, gnu::flatten]]
+		static DSOCacheEntry* find_dso_cache_entry_common (hash_t hash) noexcept
 		{
+			DSOCacheEntry *arr;
+			size_t arr_size;
+
+			if constexpr (AotCache) {
+				log_debug (LOG_ASSEMBLY, "Looking for hash 0x%x in AOT cache", hash);
+				arr = aot_dso_cache;
+				arr_size = application_config.number_of_aot_cache_entries;
+			} else {
+				log_debug (LOG_ASSEMBLY, "Looking for hash 0x%x in DSO cache", hash);
+				arr = dso_cache;
+				arr_size = application_config.number_of_dso_cache_entries;
+			}
+
 			auto equal = [](DSOCacheEntry const& entry, hash_t key) -> bool { return entry.hash == key; };
 			auto less_than = [](DSOCacheEntry const& entry, hash_t key) -> bool { return entry.hash < key; };
-			ssize_t idx = Search::binary_search<DSOCacheEntry, equal, less_than> (hash, dso_cache, application_config.number_of_dso_cache_entries);
+			ssize_t idx = Search::binary_search<DSOCacheEntry, equal, less_than> (hash, arr, arr_size);
+
 			if (idx >= 0) {
-				return &dso_cache[idx];
+				return &arr[idx];
 			}
 
 			return nullptr;
 		}
 
+		[[gnu::always_inline, gnu::flatten]]
+		static DSOCacheEntry* find_only_aot_cache_entry (hash_t hash) noexcept
+		{
+			constexpr bool IsAotCache = true;
+			return find_dso_cache_entry_common<IsAotCache> (hash);
```

Comment:
I'm not fond of a recurring pattern in this PR:

```cpp
constexpr bool IsAotCache = true;
return find_dso_cache_entry_common<IsAotCache> (hash);
```

because it's *hiding* important differences.  `find_only_aot_cache_entry()` and `find_only_dso_cache_entry()` *both* contain the expression `find_dso_cache_entry_common<IsAotCache>(hash)`, but they do "opposite" things, and that isn't clear from the callsite, as it requires "additional context" in the form of the preceding line.

I think it would be clearer if we did one of two things instead:

  * "Manually" inline, e.g. `find_dso_cache_entry_common<true>(hash)` vs. `find_dso_cache_entry_common<false>(hash)` makes it *immediately* obvious that they're different, or
  * Use a name that makes it clearer what is going on:
    ```cpp
    constexpr bool UseAotCache = true;
    return find_dso_cache_entry_common<UseAotCache> (hash);
    // vs
    constexpr bool DoNotUseAotCache = false;
    return find_dso_cache_entry_common<DoNotUseAotCache> (hash);
    ```

Different things should look different at a glance.

Reply:
I use this pattern to make it obvious what the `<true>` in template instantiation means and the name of the constant reflects the role of the template parameter in question.  What you suggest:

```c++
constexpr bool DoNotUseAotCache = false;
```

Reads to me that the call is **not** using the AOT cache (if I just look at the name of the constant) but if I notice its value, then it's confusing - because `!DoNotUseAotCache` logically means `use AOT cache`, and that's not what the call would do.  `IsAotCache = true` or `IsAotCache = false`, OTOH, is less confusing in that the name + the value are "in agreement".

I agree that the whole pattern is not perfect, but it is trying to provide for "self documenting" code or, at least, something that's easier to understand without having to find the definition of `find_dso_cache_entry_common`